8_bit_pic.py

```diff
@@ -11,102 +37,352 @@
+        self.byte_code = bytearray()
+
+        for token in self.tokenize():
+            if code_bytes := token.encode(self):
+                self.byte_code.extend(code_bytes)
+
+        logger.debug(self.byte_code)
+        logger.debug(self.symbol_table)
+
         return self.byte_code
 
     def link(self):
         return self.byte_code
 
     def compile(self):
+        self.assemble()
+        self.link()
         return self.byte_code
 
 
-@dataclass
-class CPUFlags: ...
+def resolve_int_value(assembler: vm_types.GenericAssembler, value: str, mask: int) -> int:
+    """Resolve integer referene"""
+    if value in assembler.symbol_table["symbols"]:
+        # Symbol reference.
+        value = assembler.symbol_table["symbols"][value]
+    elif value in assembler.symbol_table["map"]:
+        # Address reference.
+        value = assembler.symbol_table["map"][value]
+    else:
+        # Literal.
+        pass
+    try:
+        int_value = int(value)
+    except ValueError:
+        # Try to handle hexadecimal literals with "h" suffix.
+        int_value = int(value.lower().rstrip("h"), 16)
+    int_value = int(int_value) & mask
+    return int_value
 
 
 @dataclass
-class CPURegisters: ...
-
+class PICCentralProcessingUnit(vm_types.GenericCentralProcessingUnit):
+    RAM: bytearray
+    INSTRUCTION_MAP: ClassVar[
+        MutableMapping[InstructionCodes, Callable[[Self, int], None]]
+    ] = {}
+    ic: int = 0
+    pc: int = 0
+    wreg: int = 0
 
-@dataclass
-class CentralProcessingUnit:
     @classmethod
     def register_instruction(cls, inst_code) -> vm_types.DecoratorCallable:
         def decorator(f):
+            cls.INSTRUCTION_MAP[inst_code] = f
             return f
-
         return decorator
 
     @property
     def word_in_bytes(self) -> int:
-        return 1
+        return 2
 
     @property
     def nible_in_bytes(self) -> int:
         raise NotImplementedError
 
     def fetch(self):
-        raise NotImplementedError
+        next_ip = self.pc + self.word_in_bytes
+        self.ic = int.from_bytes(self.RAM[self.pc: next_ip])
+        self.pc = next_ip
 
     def reset(self):
-        return
+        self.ic = 0
+        self.pc = 0
+        self.wreg = 0
 
     def exec(self):
-        raise NotImplementedError
+        if self.ic & 0xff80 == 0:
+            ic = self.ic
+        elif self.ic & 0xff80 == 0x0080:
+            ic = 0x0080
+        elif self.ic & 0xff00 == 0:
+            ic = self.ic & 0xff00
+        elif self.ic & 0xf000 == 0x1000:
+            ic = self.ic & 0xfc00
+        elif self.ic & 0xf000 == 0x2000:
+            ic = self.ic & 0xf800
+        elif self.ic & 0xf000 == 0x3000:
+            ic = self.ic & 0xff00
+        else:
+            raise Exception(f"Unexpected instruction code 0x{self.ic:02x}.")
+        ic_key = InstructionCodes(ic)
+        logger.debug(f"Executing {ic_key} (0x{self.ic:04x}))...")
+        inst_func = self.INSTRUCTION_MAP[ic_key]
+        inst_func(self, self.ic)
```

Comment:
With PIC MCUs, the opcode location varies and is mixed with arguments :(

Reply:
💀 of course lol